File: internal/docker/docker.go
// Package docker shows what the Docker daemon runs next to the projects. It
// keeps one connection for the whole cockpit: the container list is fetched
// once, then the daemon's event stream keeps the cache fresh, so a hundred
// projects cost one list call and not a compose invocation each. Containers
// are matched to a project through the compose labels, and the join key is
// the compose working directory, not the compose project name, because the
// name is a normalised folder name and two folders may share it.
//
// A machine without a reachable daemon is a normal state, not an error: the
// cache answers empty, the surfaces leave the docker parts out, and the
// watcher quietly keeps trying.
package docker

import (
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

// Container is one entry of the cached list, reduced to what the surfaces
// show and the join needs.
type Container struct {
	ID      string
	Name    string
	Image   string
	State   string
	Status  string
	Health  string
	Project string
	Service string
	// WorkingDir is the compose working directory label, the join key to a
	// cockpit project. Empty for containers compose did not start.
	WorkingDir string
	Ports      []Port
	// Labels is what the daemon reported about the container, kept whole
	// because a reverse proxy in front of it writes its address in here and
	// which label that is, is configuration (see links.go). It rides along
	// with the list, so reading it costs no call of its own.
	Labels  map[string]string
	Created int64
}

// Port is one published port mapping.
type Port struct {
	Public  int
	Private int
	// Proto is empty for tcp, the unspoken default everywhere.
	Proto string
}

// Label writes the mapping the way docker ps does, public first.
func (p Port) Label() string {
	label := strconv.Itoa(p.Public) + ":" + strconv.Itoa(p.Private)
	if p.Proto != "" {
		label += "/" + p.Proto
	}
	return label
}

// Running reports whether the container is up, the one distinction the
// status color makes.
func (c Container) Running() bool {
	return c.State == "running" || c.State == "restarting" || c.State == "paused"
}

// Unwell reports the states worth an error color: a failing healthcheck or a
// daemon that gave up on the container.
func (c Container) Unwell() bool {
	return c.Health == "unhealthy" || c.State == "dead"
}

// DisplayName is what the chip shows: the compose service, the name a person
// gave the thing, with the container name as the fallback.
func (c Container) DisplayName() string {
	if c.Service != "" {
		return c.Service
	}
	return c.Name
}

// PortsLabel joins the published ports for the status line.
func (c Container) PortsLabel() string {
	labels := make([]string, len(c.Ports))
	for i, port := range c.Ports {
		labels[i] = port.Label()
	}
	return strings.Join(labels, ", ")
}

// Link is one address a browser can open for a container. The two ways a
// container is reachable share the shape:
//
// A published port is the address this page was reached on, at that port:
// Host is empty, which the client reads as its own location, and the scheme
// follows the container side of the mapping, a service listening on 443
// speaks TLS.
//
// A route a link rule read out of the labels is a host of its own, with the
// path it is routed under when it has one, and usually no port at all: the
// proxy answers on the scheme default. Its Scheme is whatever the rule pins,
// and empty, the usual answer, means the scheme of the page the link is
// opened from. Only the browser knows that one: what terminates TLS may sit
// above the proxy, where neither a label nor this server can see it.
type Link struct {
	Scheme string
	Host   string
	Port   int
	Path   string
}

// Address is the link as a person reads it, which is what the rule preview on
// the settings page shows: a route is its host and path, a published port is
// the port alone, and a scheme only appears where the link pins one.
func (l Link) Address() string {
	if l.Host == "" {
		return ":" + strconv.Itoa(l.Port)
	}
	address := l.Host
	if l.Port != 0 {
		address += ":" + strconv.Itoa(l.Port)
	}
	address += l.Path
	if l.Scheme != "" {
		return l.Scheme + "://" + address
	}
	return address
}

// State is one reading of the cache: whether a daemon answers, which host it
// is, and what it runs.
type State struct {
	Available  bool
	Host       string
	Containers []Container
}

// ForDir answers the containers whose compose working directory lies in dir,
// the per project view of the cache. The comparison also tries the resolved
// path, because compose records the path the user stood in, which may reach
// the same directory through a symlink.
func (s State) ForDir(dir string) []Container {
	if dir == "" {
		return nil
	}
	keys := []string{filepath.Clean(dir)}
	if resolved, err := filepath.EvalSymlinks(dir); err == nil {
		resolved = filepath.Clean(resolved)
		if resolved != keys[0] {
			keys = append(keys, resolved)
		}
	}
	var out []Container
	for _, c := range s.Containers {
		if c.WorkingDir == "" {
			continue
		}
		for _, key := range keys {
			if c.WorkingDir == key || strings.HasPrefix(c.WorkingDir, key+string(filepath.Separator)) {
				out = append(out, c)
				break
			}
		}
	}
	return out
}

// composeFileNames are the file names the compose CLI looks up, in its order.
var composeFileNames = []string{"compose.yaml", "compose.yml", "docker-compose.yml", "docker-compose.yaml"}

// ComposeFile answers the compose file a directory carries, the way the
// compose CLI finds it.
func ComposeFile(dir string) (string, bool) {
	for _, name := range composeFileNames {
		path := filepath.Join(dir, name)
		if info, err := os.Stat(path); err == nil && info.Mode().IsRegular() {
			return path, true
		}
	}
	return "", false
}

// Stack is one compose control point of a project: a directory compose runs
// in, with what currently runs out of it. Dir carries the project relative
// label ("" for the project root) next to the absolute path.
type Stack struct {
	Dir   string
	Label string
	// Project is the compose project name the daemon reports for the
	// containers of this directory, empty while none runs. It is what a
	// command has to name explicitly to reach exactly this stack: compose
	// otherwise derives the name from the directory it stands in.
	Project string
	Running int
	Total   int
}

// StacksForDir answers the compose control points inside dir: every distinct
// compose working directory of the project's containers, plus the project
// root while it carries a compose file of its own. A project without any of
// that answers none, which is the quiet everywhere rule.
func (s State) StacksForDir(dir string) []Stack {
	if dir == "" {
		return nil
	}
	byDir := map[string]*Stack{}
	var order []string
	for _, c := range s.ForDir(dir) {
		stack, ok := byDir[c.WorkingDir]
		if !ok {
			stack = &Stack{Dir: c.WorkingDir}
			byDir[c.WorkingDir] = stack
			order = append(order, c.WorkingDir)
		}
		stack.Total++
		if stack.Project == "" {
			stack.Project = c.Project
		}
		if c.Running() {
			stack.Running++
		}
	}
	root := filepath.Clean(dir)
	if _, ok := ComposeFile(root); ok {
		if _, exists := byDir[root]; !exists {
			byDir[root] = &Stack{Dir: root}
			order = append(order, root)
		}
	}
	sort.Strings(order)
	out := make([]Stack, 0, len(order))
	for _, key := range order {
		stack := *byDir[key]
		if rel, err := filepath.Rel(root, stack.Dir); err == nil && rel != "." {
			stack.Label = rel
		}
		out = append(out, stack)
	}
	return out
}

// sortContainers keeps the list in one stable order, compose project first,
// then service, then name, so a refresh never reshuffles the chips.
func sortContainers(list []Container) {
	sort.SliceStable(list, func(i, j int) bool {
		a, b := list[i], list[j]
		if a.Project != b.Project {
			return a.Project < b.Project
		}
		if a.Service != b.Service {
			return a.Service < b.Service
		}
		return a.Name < b.Name
	})
}

File: internal/web/static/js/dc/docker.js
import { confirm } from "@dc/dialog";
import { ensureOk, postForm } from "@dc/http";
import { notifyError, notifySuccess } from "@dc/toast";

export function navigate(url) {
  if (window.app?.navigate) window.app.navigate(url);
  else window.location.href = url;
}

export async function lifecycleAction(id, action, name) {
  try {
    const response = await postForm(`/docker/${id}/${action}`, {});
    await ensureOk(response, `Could not ${action} container "${name}".`);
    return true;
  } catch (error) {
    notifyError(error.message);
    return false;
  }
}

export async function openShell(id, kind, name) {
  try {
    const response = await postForm(`/docker/${id}/${kind}`, {});
    await ensureOk(response, `Could not open a shell for "${name}".`);
    return await response.json();
  } catch (error) {
    notifyError(error.message);
    return null;
  }
}

// composeLogs opens a terminal following a whole stack, every service of it in
// one stream. Logs are a terminal everywhere, never a dialog: a dialog holds a
// dead copy of a few hundred lines, a terminal keeps talking and is a tab like
// any other.
export async function composeLogs(project, stack, name) {
  try {
    const response = await postForm(
      `/projects/${encodeURIComponent(project)}/docker/logs`,
      { stack: stack || "" },
    );
    await ensureOk(response, `Could not open the logs of "${name || project}".`);
    return await response.json();
  } catch (error) {
    notifyError(error.message);
    return null;
  }
}

export async function compose(project, stack, action) {
  const label = action.label || action.id;
  if (action.confirm) {
    const ok = await confirm({
      title: `Run "${label}"?`,
      text: `${action.command} in ${stack ? stack : project}`,
      confirmText: "Run",
    });
    if (!ok) return false;
  }
  try {
    const response = await postForm(
      `/projects/${encodeURIComponent(project)}/docker/compose`,
      { stack: stack || "", action: action.id },
    );
    await ensureOk(response, `Could not start "${label}".`);
    notifySuccess(`${label} started.`);
    return true;
  } catch (error) {
    notifyError(error.message);
    return false;
  }
}

export async function restoreActions() {
  try {
    const response = await postForm("/docker/actions/restore", {});
    await ensureOk(response, "Could not restore the default actions.");
    notifySuccess("The default compose actions are back.");
    return true;
  } catch (error) {
    notifyError(error.message);
    return false;
  }
}

// restoreLinkRules is the same way back for the rules that read a container's
// own address out of its labels: one route, and it clears the setting instead
// of storing today's defaults into it.
export async function restoreLinkRules() {
  try {
    const response = await postForm("/docker/link-rules/restore", {});
    await ensureOk(response, "Could not restore the default link rules.");
    notifySuccess("The default link rules are back.");
    return true;
  } catch (error) {
    notifyError(error.message);
    return false;
  }
}

// linkUrl builds what a link opens. Only the browser knows two of the three
// parts: a published port is reached on the host this page was reached on, and
// a proxy route that names no scheme of its own is reached over the scheme
// this page was reached over, so it is opened protocol relative. The server
// cannot answer either one, its own X-Forwarded-Proto is what the proxy's
// entrypoint is, not what the person in front of it typed.
function linkUrl(link) {
  const host = link.host || window.location.hostname;
  const port = link.port ? `:${link.port}` : "";
  const prefix = link.scheme ? `${link.scheme}:` : "";
  return `${prefix}//${host}${port}${link.path || ""}`;
}

// linkAddress is the address as a person reads it, the same shape the settings
// preview shows: a route is its host and path, a published port is the port.
function linkAddress(link) {
  return link.host ? `${link.host}${link.path || ""}` : `:${link.port}`;
}

// LABEL_TAIL caps how much of an address never shrinks. A path can be long,
// and a tail that takes the whole row would starve the head it is supposed to
// complete.
const LABEL_TAIL = 24;

// addressLabel splits an address so the ellipsis lands in the middle. What
// tells the sibling hosts of one container apart is their end, so the end is
// the part that must survive: the tail starts at the last dot of the host and runs
// to the end, path included, and the head takes whatever is left. An address
// whose tail alone would be too long is cut a fixed distance from its end
// instead, which keeps the same promise.
function addressLabel(address) {
  const found = /\.[^./]+(\/.*)?$/.exec(address);
  let cut = found ? found.index : -1;
  if (cut < 0 || address.length - cut > LABEL_TAIL) cut = Math.max(0, address.length - LABEL_TAIL);
  return { head: `Open ${address.slice(0, cut)}`, tail: address.slice(cut) };
}

// linkItems are the addresses one container answers on, as menu entries: the
// routes first, where one exists it is the address a person wants, then a
// divider and the published ports in their numeric order. The same address
// twice is one entry.
function linkItems(links) {
  const seen = new Set();
  const routes = [];
  const ports = [];
  for (const link of links || []) {
    const key = `${link.scheme || ""}//${link.host || ""}:${link.port || 0}${link.path || ""}`;
    if (seen.has(key)) continue;
    seen.add(key);
    (link.host ? routes : ports).push(link);
  }
  ports.sort((a, b) => a.port - b.port);
  const items = routes.map(linkItem);
  if (routes.length && ports.length) items.push({ divider: true });
  return items.concat(ports.map(linkItem));
}

function linkItem(link) {
  const address = linkAddress(link);
  return {
    label: addressLabel(address),
    title: `Open ${address}`,
    icon: "ti-external-link",
    action: () => window.open(linkUrl(link), "_blank", "noopener"),
  };
}

export function containerMenuItems(info, { onShell } = {}) {
  const items = [];
  // No status line: what the daemon calls the uptime is a snapshot of the last
  // cache refresh and ages badly, and the icon color already says running,
  // stopped or unwell. What the ports are is worth the line.
  if (info.portsLabel) {
    items.push({ label: info.portsLabel, icon: "ti-plug", disabled: true });
    items.push({ divider: true });
  }
  items.push(...linkItems(info.links));
  if (info.running && info.cli !== false && onShell) {
    items.push({ label: "Shell", icon: "ti-terminal-2", action: () => void onShell(info, "shell") });
    items.push({ label: "Logs", icon: "ti-file-text", action: () => void onShell(info, "logs-shell") });
  }
  items.push({ divider: true });
  if (info.running) {
    items.push({ label: "Restart", icon: "ti-refresh", warn: true, action: () => void lifecycleAction(info.id, "restart", info.name) });
    items.push({ label: "Stop", icon: "ti-player-stop", warn: true, action: () => void lifecycleAction(info.id, "stop", info.name) });
  } else {
    items.push({ label: "Start", icon: "ti-player-play", action: () => void lifecycleAction(info.id, "start", info.name) });
  }
  return items;
}

// projectMenuItems is the project's own docker menu, the same list on the
// projects page and in the editor: where to reach what runs, then per stack
// its logs, its last run and the configured commands.
//
// **The project menu answers which container, the container's own menu answers
// which address.** A project can run a dozen containers and one of them can be
// routed under a dozen host names; all of them in one list is a wall in front
// of the one entry somebody wants. So a container gets at most one entry here.
// With exactly one address that entry is the address itself, the most
// informative thing it can say. With several it names the container and how
// many there are, and opening it opens the same menu again with that
// container's addresses and a way back (onDrill). Picking the first of a
// dozen hosts for somebody is precisely what the cockpit cannot know.
export function projectMenuItems({ project, containers = [], stacks = [], actions = [], onLogs, onDrill } = {}) {
  const items = [];
  containers.forEach((container) => {
    const links = container.links || [];
    if (!links.length) return;
    if (links.length === 1) {
      items.push(linkItem(links[0]));
      return;
    }
    // A surface that cannot drill in lists them rather than picking one.
    if (!onDrill) {
      items.push(...linkItems(links));
      return;
    }
    items.push({
      label: `${container.name} (${links.length} addresses)`,
      title: `${links.length} addresses of ${container.name}`,
      icon: "ti-external-link",
      action: () => onDrill([
        { label: "Back", icon: "ti-arrow-left", action: () => onDrill(null) },
        { divider: true },
        ...linkItems(links),
      ]),
    });
  });
  stacks.forEach((stack) => {
    if (items.length) items.push({ divider: true });
    items.push(...stackMenuItems(stack, project, actions, onLogs));
  });
  return items;
}

function stackMenuItems(stack, project, actions, onLogs) {
  const items = [];
  const suffix = stack.label ? ` (${stack.label})` : "";
  items.push({
    label: (stack.total ? `${stack.running} of ${stack.total} running` : "Nothing running") + suffix,
    icon: "ti-stack-2",
    disabled: true,
  });
  if (stack.total && onLogs) {
    items.push({ label: `Logs${suffix}`, icon: "ti-file-text", action: () => void onLogs(stack) });
  }
  if (stack.run) {
    items.push({
      label: stack.run.running ? `${stack.run.action} is running…` : `Output of ${stack.run.action}`,
      icon: stack.run.running ? "ti-loader-2" : "ti-file-description",
      iconClass: stack.run.running ? "dc-spin" : "",
      action: () => navigate(stack.run.url),
    });
  }
  if (stack.busy) return items;
  if (!actions.length) {
    items.push({ label: "No compose actions configured.", icon: "ti-plug-off", disabled: true });
    items.push({ label: "Restore the default actions", icon: "ti-restore", action: () => void restoreActions() });
    return items;
  }
  actions.forEach((action) => {
    items.push({
      label: `${action.label}${suffix}`,
      icon: action.icon,
      warn: Boolean(action.confirm),
      action: () => void compose(project, stack.label, action),
    });
  });
  return items;
}
